### ISSUE TITLE
onError should be called for any error

### ISSUE THREAD

**Issue by janhancic** [redux-saga/redux-saga#1308]
I was writing some tests and using `runSaga` with the `onError` handler. I was passing a fake error which was just an empty object. And my onError handler didn't get called. After digging through the source code I found that you're doing a typeof check:

https://github.com/redux-saga/redux-saga/blob/0d409bd2344032ca09a7c6962c56d28ea0917d44/src/internal/proc.js#L300

I don't think you need to do that, it's perfectly valid to throw an error that does not inherit from `Error`.

**Comment by Andarist** [comment]
Thanks for raising this, Im evaluating error handling at the moment and I might change this soon-ish.

**Comment by ggarek** [comment]
@Andarist , thank you for your work on the project, it is great! And there pre-release changes has many things i wanted to see. Thank you!
 
i was checking the latest prerelease (v1.0.0.beta.1). The description says

> Also onError should get called now even if you throw non-Errors in your code.

But i think it is not the case, still only `instance of Error` is reported.. Would be happy to have this feature 🙏

**Comment by Andarist** [comment]
Hm, looking into the source code each "error" [should get reported](https://github.com/redux-saga/redux-saga/blob/v1.0.0-beta.1/packages/core/src/internal/proc.js#L346). Could u double check that you are using v1.0.0-beta.1?

**Comment by ggarek** [comment]
I have rechecked my test example. Here are the details

### TL;DR
- I compare two version of redux saga: `0.16.0` and `1.0.0-beta.1`
- I test for two things: 
  1) i expect `put` effect to rethrow errors _(ℹ️ i know, there is `put.resolve`)_
  2) i expect `onError` to get called for uncaught errors thrown in saga _(in fact, in put effect)_

##### Observations
||`put` propagates `Errors`|`put` propagates `not Errors`|`onError` is called for uncaught `Errors`|`onError` is called for uncaught `not Errors`|
|-|-|-|-|-|
|v0.16.0|🚫|🚫|- _(because put swallows)_|- _(because put swallows)_|
|v1.0.0-beta.1| ✅ |✅|✅|🚫|

### Details

#### Here is my test

```js
import { applyMiddleware, createStore } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { put, takeEvery } from "redux-saga/effects";

// TODO: Action handled by saga is passed to reducer first.. is it breaking change? Or it was always like that?

describe('sagas/vendor', () => {
  describe('put effect should not swallow', () => {
    const sequence = [];
    const errorThrowInReducer = new Error('an instance of Error');
    const notErrorThrownInReducer = 'NOT Error, just a String';

    // create reducer which throws
    const reducer = (state = null, action: any) => {
      switch(action.type) {
        case 'PLEASE_THROW_ERROR':
          sequence.push('reducer: throw');
          throw errorThrowInReducer;
        case 'PLEASE_THROW_NOT_ERROR':
          sequence.push('reducer: throw');
          throw notErrorThrownInReducer;
        default:
          return state;
      }
    }

    // create root saga and saga to put to the reducer
    function* oneSaga(action) {
      sequence.push('oneSaga: start');
      if (action.type === 'SAGA_PLEASE_THROW_NOT_ERROR') {
        yield put({ type: 'PLEASE_THROW_NOT_ERROR' });
      } else {
        yield put({ type: 'PLEASE_THROW_ERROR' });
      }
    }
    function* rootSaga() {
      yield takeEvery(['SAGA_PLEASE_THROW_ERROR', 'SAGA_PLEASE_THROW_NOT_ERROR'], oneSaga);
    }
    const safely = (originalSaga: any) => function* _safely() {
      try {
        yield originalSaga();
      } catch(e) {
        sequence.push('safely: catch error', e);

        // Rethrow so we can check onError
        throw e;
      }
    }

    // create and run saga middleware
    const onError = (e) => {
      sequence.push('onError: called', e);
    };
    const createNewStore = () => {
      const sagaMiddleware = createSagaMiddleware({
        onError
      });

      // create store with the reducer
      const store = createStore(reducer, applyMiddleware(sagaMiddleware));
      sagaMiddleware.run(safely(rootSaga));
      return store;
    }


    // dispatch action to cause error
    // assert onError was called and error was propagated
    it('should call, not swallow, if reducer throws Error', () => {
      expect(() => {
        createNewStore().dispatch({ type: 'SAGA_PLEASE_THROW_ERROR' });
      }).not.toThrow();
      expect(sequence).toEqual([
        'oneSaga: start',
        'reducer: throw',
        'safely: catch error',
        errorThrowInReducer,
        'onError: called',
        errorThrowInReducer,
      ]);
    });

    it('should call, not swallow, if reducer throws not Error', () => {
      sequence.length = 0;
      expect(() => {
        createNewStore().dispatch({ type: 'SAGA_PLEASE_THROW_NOT_ERROR' });
      }).not.toThrow();
      expect(sequence).toEqual([
        'oneSaga: start',
        'reducer: throw',
        'safely: catch error',
        notErrorThrownInReducer,
        'onError: called',
        notErrorThrownInReducer
      ]);
    });
  });
});
```

#### v0.16.0 results

``` bash
> NODE_ENV=test jest "vendor.saga.test.ts"

 FAIL  ./vendor.saga.test.ts
  sagas/vendor
    put effect should not swallow
      ✕ should call not swallow if reducer throws Error (25ms)
      ✕ should call not swallow if reducer throws not Error (2ms)

  ● sagas/vendor › put effect should not swallow › should call not swallow if reducer throws Error

    expect(received).toEqual(expected)

    Expected value to equal:
      ["oneSaga: start", "reducer: throw", "safely: catch error", [Error: an instance of Error], "onError: called", [Error: an instance of Error]]
    Received:
      ["oneSaga: start", "reducer: throw"]

    Difference:

    - Expected
    + Received

      Array [
        "oneSaga: start",
        "reducer: throw",
    -   "safely: catch error",
    -   [Error: an instance of Error],
    -   "onError: called",
    -   [Error: an instance of Error],
      ]

      70 |         createNewStore().dispatch({ type: 'SAGA_PLEASE_THROW_ERROR' });
      71 |       }).not.toThrow();
    > 72 |       expect(sequence).toEqual([
      73 |         'oneSaga: start',
      74 |         'reducer: throw',
      75 |         'safely: catch error',

      at Object.<anonymous> (vendor.saga.test.ts:72:24)

  ● sagas/vendor › put effect should not swallow › should call not swallow if reducer throws not Error

    expect(received).toEqual(expected)

    Expected value to equal:
      ["oneSaga: start", "reducer: throw", "safely: catch error", "NOT Error, just a String", "onError: called", "NOT Error, just a String"]
    Received:
      ["oneSaga: start", "reducer: throw"]

    Difference:

    - Expected
    + Received

      Array [
        "oneSaga: start",
        "reducer: throw",
    -   "safely: catch error",
    -   "NOT Error, just a String",
    -   "onError: called",
    -   "NOT Error, just a String",
      ]

      86 |         createNewStore().dispatch({ type: 'SAGA_PLEASE_THROW_NOT_ERROR' });
      87 |       }).not.toThrow();
    > 88 |       expect(sequence).toEqual([
      89 |         'oneSaga: start',
      90 |         'reducer: throw',
      91 |         'safely: catch error',

      at Object.<anonymous> (vendor.saga.test.ts:88:24)

  console.error node_modules/redux-saga/lib/internal/utils.js:240
    uncaught at oneSaga Error: an instance of Error
        at Suite.<...>
        at process._tickCallback (internal/process/next_tick.js:188:7)

  console.error node_modules/redux-saga/lib/internal/utils.js:240
    uncaught at oneSaga NOT Error, just a String

Test Suites: 1 failed, 1 total
Tests:       2 failed, 2 total
Snapshots:   0 total
Time:        1.78s
Ran all test suites matching /vendor.saga.test.ts/i.
npm ERR! Test failed.  See above for more details.
```

#### v1.0.0-beta.1 results

```
> NODE_ENV=test jest "vendor.saga.test.ts"

 FAIL  ./vendor.saga.test.ts
  sagas/vendor
    put effect should not swallow
      ✓ should call not swallow if reducer throws Error (9ms)
      ✕ should call not swallow if reducer throws not Error (12ms)

  ● sagas/vendor › put effect should not swallow › should call not swallow if reducer throws not Error

    expect(received).toEqual(expected)

    Expected value to equal:
      ["oneSaga: start", "reducer: throw", "safely: catch error", "NOT Error, just a String", "onError: called", "NOT Error, just a String"]
    Received:
      ["oneSaga: start", "reducer: throw", "safely: catch error", "NOT Error, just a String"]

    Difference:

    - Expected
    + Received

      Array [
        "oneSaga: start",
        "reducer: throw",
        "safely: catch error",
        "NOT Error, just a String",
    -   "onError: called",
    -   "NOT Error, just a String",
      ]

      86 |         createNewStore().dispatch({ type: 'SAGA_PLEASE_THROW_NOT_ERROR' });
      87 |       }).not.toThrow();
    > 88 |       expect(sequence).toEqual([
      89 |         'oneSaga: start',
      90 |         'reducer: throw',
      91 |         'safely: catch error',

      at Object.<anonymous> (vendor.saga.test.ts:88:24)

  console.error node_modules/redux-saga/lib/internal/utils.js:300
    NOT Error, just a String

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 passed, 2 total
Snapshots:   0 total
Time:        1.891s
Ran all test suites matching /vendor.saga.test.ts/i.
npm ERR! Test failed.  See above for more details.
```

**Comment by eliseumds** [comment]
The version that was published to NPM still includes the logic that checks where error is an instance of of Error. I extracted this from my *node_modules* with v1.0.0-beta.1:

```js
// es/internal/proc.js:372

if (result instanceof Error && onError) {
```

See the released version at https://github.com/redux-saga/redux-saga/compare/v1.0.0-beta.1...master. It does not include commit https://github.com/redux-saga/redux-saga/commit/7da560.

**Comment by eliseumds** [comment]
The version that was published to NPM still includes the logic that checks where error is an instance of of Error. I extracted this from my *node_modules* with v1.0.0-beta.1:

```js
// es/internal/proc.js:372

if (result instanceof Error && onError) {
```

It feels like it wasn't rebuilt.

**Comment by restrry** [comment]
that instanceOf check should have been removed https://github.com/redux-saga/redux-saga/pull/1398
and `beta-v1` contains those changes https://github.com/redux-saga/redux-saga/blob/v1.0.0-beta.1/packages/core/src/internal/proc.js#L345-L346
anyway I've just added test case for it https://github.com/redux-saga/redux-saga/pull/1517

**Comment by Andarist** [comment]
It seems that it really could have not been rebuild. We are probably going to prepare a new release in following weak or two - I'll make sure to check our build process if it is setup correctly for publication.

**Comment by Andarist** [comment]
The new release is out for some item, we should also release final v1 this month. Hope this resolves your problem, closing this issue for now.
